packages/malloy-render/src/component/render-result-metadata.ts

```diff
@@ -21,283 +22,1562 @@
  * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
-import {
-  DataArray,
-  DataColumn,
-  Explore,
-  ExploreField,
-  Field,
-  QueryData,
-  Result,
-} from '@malloydata/malloy';
 import {Tag} from '@malloydata/malloy-tag';
-import {
-  getFieldKey,
-  valueIsDateTime,
-  valueIsNumber,
-  valueIsBoolean,
-  valueIsString,
-} from './util';
-import {
-  DataRowWithRecord,
-  RenderResultMetadata,
-  VegaChartProps,
-  VegaConfigHandler,
-} from './types';
-import {
-  NULL_SYMBOL,
-  shouldRenderAs,
-  shouldRenderChartAs,
-} from './apply-renderer';
+import {NULL_SYMBOL, tagFromAnnotations, valueToMalloy} from './util';
+import {DrillEntry, VegaChartProps, VegaConfigHandler} from './types';
 import {mergeVegaConfigs} from './vega/merge-vega-configs';
 import {baseVegaConfig} from './vega/base-vega-config';
 import {renderTimeString} from './render-time';
 import {generateBarChartVegaSpec} from './bar-chart/generate-bar_chart-vega-spec';
-import {createResultStore} from './result-store/result-store';
+import {createResultStore, ResultStore} from './result-store/result-store';
 import {generateLineChartVegaSpec} from './line-chart/generate-line_chart-vega-spec';
-import {parse, Config} from 'vega';
-
-function createDataCache() {
-  const dataCache = new WeakMap<DataColumn, QueryData>();
-  return {
-    get: (cell: DataColumn) => {
-      if (!dataCache.has(cell) && cell.isArray()) {
-        const data: DataRowWithRecord[] = [];
-        const fields = cell.field.allFields;
-        for (const row of cell) {
-          const record = {__malloyDataRecord: row};
-          for (const field of fields) {
-            const value = row.cell(field).value;
-            // TODO: can we store Date objects as is? Downstream chart code would need to be updated
-            record[field.name] =
-              value instanceof Date ? value.valueOf() : row.cell(field).value;
-          }
-          data.push(record as DataRowWithRecord);
-        }
-        dataCache.set(cell, data);
-      }
-      return dataCache.get(cell)!;
-    },
-  };
-}
+import {parse, Config, Runtime} from 'vega';
+import * as Malloy from '@malloydata/malloy-interfaces';
 
 export type GetResultMetadataOptions = {
   getVegaConfigOverride?: VegaConfigHandler;
 };
 
 export function getResultMetadata(
-  result: Result,
+  result: Malloy.Result,
   options: GetResultMetadataOptions = {}
 ) {
-  const fieldKeyMap: WeakMap<Field | Explore, string> = new WeakMap();
-  const getCachedFieldKey = (f: Field | Explore) => {
-    if (fieldKeyMap.has(f)) return fieldKeyMap.get(f)!;
-    const fieldKey = getFieldKey(f);
-    fieldKeyMap.set(f, fieldKey);
-    return fieldKey;
-  };
+  const fields: Malloy.DimensionInfo[] = [];
+  for (const field of result.schema.fields) {
+    if (field.kind === 'dimension') {
```

Comment:
Makes me wonder whether a `Result` type should just have a single root field already created for the result...